x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -131,98 +127,83 @@ private void createFollowerIndex(
-                updatedState = allocationService.reroute(
-                        ClusterState.builder(updatedState).routingTable(routingTableBuilder.build()).build(),
-                        "follow index [" + request.getFollowRequest().getFollowerIndex() + "] created");
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        };
+
+        Settings.Builder settingsBuilder = Settings.builder()
+            .put(IndexMetaData.SETTING_INDEX_PROVIDED_NAME, request.getFollowRequest().getFollowerIndex())
+            .put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true);
+        String leaderClusterRepoName = CcrRepository.NAME_PREFIX + remoteCluster;
+        RestoreService.RestoreRequest restoreRequest = new RestoreService.RestoreRequest(leaderClusterRepoName,
+            CcrRepository.LATEST, new String[]{request.getLeaderIndex()}, request.indicesOptions(),
+            "^(.*)$", request.getFollowRequest().getFollowerIndex(), Settings.EMPTY, request.masterNodeTimeout(), false,
+            false, true, settingsBuilder.build(), new String[0],
+            "restore_snapshot[" + leaderClusterRepoName + ":" + request.getLeaderIndex() + "]");
+        initiateRestore(restoreRequest, restoreCompleteHandler);
+    }
 
-                logger.info("[{}] creating index, cause [ccr_create_and_follow], shards [{}]/[{}]",
-                        followIndex, followIMD.getNumberOfShards(), followIMD.getNumberOfReplicas());
+    private void initiateRestore(RestoreService.RestoreRequest restoreRequest, ActionListener<RestoreSnapshotResponse> listener) {
+        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(() -> {
+            try {
+                restoreService.restoreSnapshot(restoreRequest, new ActionListener<RestoreService.RestoreCompletionResponse>() {
+                    @Override
+                    public void onResponse(RestoreService.RestoreCompletionResponse restoreCompletionResponse) {
+                        final Snapshot snapshot = restoreCompletionResponse.getSnapshot();
+                        final String uuid = restoreCompletionResponse.getUuid();
+                        clusterService.addListener(new RestoreClusterStateListener(clusterService, uuid, snapshot, listener));
+                    }
 
-                return updatedState;
+                    @Override
+                    public void onFailure(Exception e) {
+                        listener.onFailure(e);
+                    }
+                });
+            } catch (Exception e) {
+                listener.onFailure(e);
             }
         });
     }
 
     private void initiateFollowing(
-            final PutFollowAction.Request request,
-            final ActionListener<PutFollowAction.Response> listener) {
+        final Client client,
+        final PutFollowAction.Request request,
+        final ActionListener<PutFollowAction.Response> listener) {
```

Comment:
We'll probably have to adapt the response listener as well here. Restoring a multi-gigabyte shard might take a while, and I'm not sure we want to wait on the response of the restore here. When restoring a snapshot, we have given the option to the user "waitForCompletion = true | false", perhaps we need the same here, defaulting to false.

Reply:
I added a parameter.